x-pack/plugins/security_solution/server/lib/detection_engine/signals/utils.ts

```diff
@@ -661,3 +663,356 @@ export const createTotalHitsFromSearchResult = ({
+};
+
+/**
+ * The basic gist is find all indices that have the given timestamp(s)
+ * If there is a timestamp override, ensure all indices for all index
+ * patterns have that timestamp override. If an index pattern does not,
+ * check if that index has an '@timestamp' field and default to that.
+ *
+ * If no timestamp override field is found, use the '@timestamp' field.
+ * If an index in a given index pattern is missing the '@timestamp' field
+ * set the status of the rule to be a partial failure
+ *
+ * If all index patterns are missing the '@timestamp' field and the override
+ * set the status to error.
+ * @param indices string[]
+ * @param timestamps string[]
+ * @param services AlertServices
+ * @param logger Logger
+ * @param buildRuleMessage
+ */
+export const timestampFieldCheck = async (
+  indices: string[],
+  timestamps: string[],
+  services: AlertServices,
+  logger: Logger,
+  buildRuleMessage: BuildRuleMessage
+): Promise<PreCheckRuleResultInterface> => {
+  const timestampsAndIndices = await checkIndexMappingsForTimestampFields(
+    indices,
+    timestamps,
+    services,
+    logger,
+    buildRuleMessage
+  );
+
+  if (isEmpty(timestampsAndIndices)) {
+    return {
+      result: 'success',
+      resultMessages: [
+        `Could not find field mappings for ${JSON.stringify(
+          timestamps
+        )} in indices: ${JSON.stringify(indices)}`,
+      ],
+      successIndexes: [],
+      failingIndexes: [],
+      timestampsAndIndices: {},
+    };
+  }
+
+  // get all indices which match the given index pattern(s)
+  // we will use this for determining which indices within
+  // a given index pattern are missing timestamp field(s)
+  // IMPORTANT: this uses cat.indices api which requires
+  // 'monitor' privilege on both the cluster and the index
+  // patterns
+  const indexPatternIndices = await getIndexesMatchingIndexPatterns(
+    indices,
+    services,
+    logger,
+    buildRuleMessage
+  );
+
+  // if timestampsAndIndices[timestampOverride].length === Object.values(indexPatternIndices).length
+  // then skip the @timestamp field since we have all of the indices in the timestamp override.
+  const timestampOverrideField = Object.keys(timestampsAndIndices).filter(
+    (timestamp) => timestamp !== '@timestamp'
+  )[0];
+
+  if (
+    !isEmpty(indexPatternIndices) &&
+    timestampOverrideField != null &&
+    Object.keys(timestampsAndIndices).includes(timestampOverrideField)
+  ) {
+    if (
+      timestampsAndIndices[timestampOverrideField].length ===
+      Object.values(indexPatternIndices).flat(2).length
+    ) {
+      delete timestampsAndIndices['@timestamp'];
+    }
+  }
+
+  const timestampKeys = Object.keys(timestampsAndIndices);
+
+  if (timestampKeys.length === 0) {
+    throw Error(`No indices contained timestamp fields: ${JSON.stringify(timestamps)}`);
+  }
+
+  const indexPatternRegEx = indices.reduce(
+    (acc, indexPattern) => ({ [indexPattern]: new RegExp(indexPattern), ...acc }),
```

Comment:
Index patterns are more like unix globs than regular expressions so we shouldn't try to make a regex directly from the index pattern.